### ISSUE TITLE
Logger.LogLevel in ESAPI.properties not working if set to ERROR/FATAL

### ISSUE THREAD

**Issue by benz-jemi** [ESAPI/esapi-java-legacy#593]
We have a requirement to hide few input data in logs. ESAPI prints log saying " [WARNING ] Multiple (2x) encoding detected in <INPUT> " as input data had multiencoded characters. 

But we don't want this input data logged. So I tried changing the log level to ERROR/FATAL in esapi.properties through Logger.LogLevel but its not working. 

Expectations/Queries:
1) It would be great if we have a configuration like Logger.LogInputData to hide the inputs from logging or programmatically for each request.
2) Why Logger.LogLevel not working? We have ESAPI 2.0.1 and later tried  this config with 2.2.0.0 too.

**Comment by kwwall** [comment]
@benz-jemi -- Can you provide a little more detail please, such as:

- What version of ESAPI are you using?
- What logger are you using (JUL or SLF4J)?
- If you are using SLF4J, could you share its configuration for whatever backing logger you are using there, such as log4j 2 or logback, etc.?

Thanks.

**Comment by kwwall** [comment]
I'm not sure if this is a bug or a "feature". `Logger.LogLevel` is only referenced by` DefaultSecurityConfiguration.getLogLevel()` and that method is not really used anywhere in the ESAPI code outside of the JUnit tests! Likewise, I found no references to `DefaultSecurityConfiguration.LOG_LEVEL` either, other than by the `getLogLevel()` method.

@benz-jemi - The bottom line, if you want to disable ESAPI logging, you will currently need to do it via the configuration of the underlying logger (JUL or whatever logging implementation you are using with SLF4J).

@jeremiahjstacey - I haven't checked the git history at all, but do you happen to know if `getLogLevel()` used to be used at all or was it, to your knowledge, never used by ESAPI? Note that `SecurityConfiguration.getLogLevel()` _is_ marked as 'deprecated' (albeit for a completely different reason). Regardless, it looks like we need to consider deprecating `SecurityConfiguration.getLogLevel()` to note that it is no longer used as well as doing the same for `DefaultSecurityConfiguration.LOG_LEVEL` and the `Logger.LogLevel` property in **ESAPI.properties**. OTOH, if the removal of honoring `Logger.LogLevel` was an accidental "feature" or restructuring the ESAPI loggers, then we need to have a discussion--at least amongst ourselves--as to whether or not we should restore that behavior or if we actually want to deprecate it and just fix our documentation.

**Comment by jeremiahjstacey** [comment]
@kwwall I remember spending some time going over the Java and Log4J implementations before structuring the SLF4J logger. I do not recall any uses of `SecurityConfiguration.getLogLevel()` references. 
Looking at the JavaLogFactory class from the 2.1 branch, there are no uses of that invocation in that implementation either.  In 2.1 JavaLogFactory is the factory, bridge, and logger classes as one unit.
https://github.com/ESAPI/esapi-java-legacy/blob/2.1/src/main/java/org/owasp/esapi/reference/JavaLogFactory.java

**Comment by kwwall** [comment]
Thanks @jeremiahjstacey. Your research caused me to look back even further. I looked in the '1.4' GitHub branch and found calls to `getLogLevel()` there in both `JavaLogFactory` and `Log4JLogFactory` classes. It was also still there in the '1.5_international' branch, but was gone in the '2.0_quality' branch, during which time the logging code was significantly restructured. It probably was dropped with that restructuring.

But it has been good for a long time, so we should delete all of it and do other clean up of the logging usage internal to ESAPI (e.g., see somewhat related GitHub issue #595) so others won't get confused about this in the future. I am working on some changes now, so I will take a look at some of this and see what I can clean up.

**Comment by benz-jemi** [comment]
@kwwall - we used esapi-2.0.1.jar and org.owasp.esapi.reference.JavaLogFactory. 

But have you read my other query, we have few services which takes in PII data and so we don't want any log of the data. Since it has multiencoded characters, Warning log was printed along with the PII data. So to resolve this, we upgraded to esapi-2.2.1.1.jar and used esapi-java-logging.properties  with the following configurations
handlers= java.util.logging.ConsoleHandler
.level= FATAL
java.util.logging.ConsoleHandler.level = FATAL
java.util.logging.ConsoleHandler.formatter = java.util.logging.SimpleFormatter
java.util.logging.SimpleFormatter.format=[%1$tF %1$tT] [%3$-7s] %5$s %n

and now that warning is not logged. However is it possible to have a configuration like Logger.LogInputData in future release or is it possible to turn off ESAPI logs(Because we use esapi encoder and catch any exception and log it ourselves) If latter is possible, Could any possible security violation scenarios be missed from our notice or what would the impact?

**Comment by kwwall** [comment]
@benz-jemi - I'm not sure I know what you mean in reference to "a configuration like `Logger.LogInputData`". Could you please provide a reference? That said, as I mentioned [here](https://github.com/ESAPI/esapi-java-legacy/issues/594#issuecomment-781038054) in issue #594, most of the _exceptions_ that are logged within ESAPI are a result of `IntrusionException` so you could try disabling the ESAPI Intrusion Detector as described there.

Lastly, in theory at least, one should be able to use the underlying native logging configuration (e.g. JUL, log4j 2 used with SLF4J, etc.) to set different log levels for different ESAPI classes. In pretty much all the classes, we do something like:
`private static Logger logger = ESAPI.getLogger("Encoder");`
or
`private static Logger logger = ESAPI.getLogger(StringValidationRule.class);`

so each logger should have a more or less unique name. I know that log4j 1 used to support this, although it has been a dozen plus years since I used it, but we used to use it to dynamically enabled DEBUG level logging for certain classes in production. (There was a thread that re-read the log4j.properties file if it was updated).  That certainly is the way that we _intended_ ESAPI to be used I think. (On the other hand, after going through a few mostly usually posts about this on Stack Overflow, I have grown weary of looking. I will trust that your persistence in pursuing an answer is more motivated than mine.)

**Comment by kwwall** [comment]
@benz-jemi - Also, one more thing... I have a test "ESAPI Hello World" example set up where I tested essentially this:
```
        Logger logger = ESAPI.log();

        logger.always(Logger.EVENT_SUCCESS, "In App.main(): ALWAYS: Hello ESAPI World!");
        logger.info(Logger.EVENT_SUCCESS, "In App.main(): INFO: Hello ESAPI World!");
        logger.warning(Logger.EVENT_SUCCESS, "In App.main(): WARNING: Hello ESAPI World!");
        logger.error(Logger.EVENT_SUCCESS, "In App.main(): ERROR: Hello ESAPI World!");
        logger.debug(Logger.EVENT_SUCCESS, "In App.main(): DEBUG: Hello ESAPI World!");
```

with my **esapi-java-logging.properties** file set up like this:
```
handlers= java.util.logging.ConsoleHandler
.level= ERROR
java.util.logging.ConsoleHandler.level = ERROR
java.util.logging.ConsoleHandler.formatter = java.util.logging.SimpleFormatter
java.util.logging.SimpleFormatter.format=[%1$tF %1$tT] [%3$-7s] %5$s %n

```

and I only see the ALWAYS and ERROR logging as expected:
```
[2021-02-22 11:43:04] [DefaultLogger] [EVENT SUCCESS Anonymous:@unknown -> /ExampleApplication/DefaultLogger] In App.main(): ALWAYS: Hello ESAPI World! 
[2021-02-22 11:43:04] [DefaultLogger] [EVENT SUCCESS Anonymous:@unknown -> /ExampleApplication/DefaultLogger] In App.main(): ERROR: Hello ESAPI World! 

```
So my first suggestion is to do something like that _directly_ in your code and make sure that your application is picking up the **esapi-java-logging.properties** file that you think it should be. (I know I had one under $HOME/.esapi that was throwing me off for a while.)

**Comment by jeremiahjstacey** [comment]
I believe that @benz-jemi is asking if a configuration exists that would allow the warning event to be logged, but have that log event exclude the String input that triggered the violation.  (effectively not capturing the PII in the log output)

If I'm understanding that question correctly, the answer to that is no.  The project configuration does not enable that granularity of control over what data is sent to the underlying logging implementation.

That feature has also not been discussed as part of the current or future roadmap (to the best of my knowledge).

**Comment by jeremiahjstacey** [comment]
**Can we turn off ESAPI logs?**
Yes.  @kwwall  has identified three different methods you can use to disable logging output from ESAPI.  You will need to evaluate each option for your use cases to determine which is the most appropriate.  I've enumerated the options I've seen presented below:
1.  Set the overall Java Logging Level to restrict the log events propagated to the logging implementation.  This is a configuration setting that will impact All of the ESAPI logging, and may also impact other tools using the Java Logging depending on how your runtime is configured.
2. Add/Extend custom package-level handlers for the org.owasp.esapi.* packaging in your java logging configuration.  This option is similar to above, except that it is isolating the log output from the ESAPI packages.  Again,  this behavior will be dependent on how you have your project configured and may require some testing to ensure you're getting the desired output from ESAPI and not impacting other areas of your application.  As a courtesy, [This SoF Link](https://stackoverflow.com/questions/3920930/java-util-logging-how-to-set-level-by-logger-package-or-prefix) can help to get you started down this path.
3.  Disable the multiple encoding checking for ESAPI in your system.   In your deployed ESAPI.properties file, set the value of `Encoder.AllowMultipleEncoding` to `true`.  This will disable the check in the system which is verifying that input has not been encoded by one or more encoders.  This has the potential impact of allowing your system to ingest input that may have security concerns.  You should evaluate the source of your PII and assert that you want to implicitly trust the input.  This, again, is highly dependent on your environment and the results you are looking for, and you should evaluate the risks/benefits to your system.

There may be impacts to each of these configurations that I am not considering, so I very strongly recommend that you evaluate them for their individual merit in each of your pertinent use cases in your specific deployed environment.

**Can we Catch the IntrusionException?**
If you are calling the Encoder directly from your baseline, then you should apply the behavior that best suits the intent of your baseline.  Depending on your configuration and baseline implementation, the three options above (or some combination thereof) may abate this need.   Functionally you can do this.  I believe it to be best to leave the determination whether you should do this up to you and your validation of the three aforementioned options.

**Additional Resources**
If all you are using the ESAPI baseline for is the Encoder options, I'd like to point out a recommendation in the [ESAPI project documentation](https://owasp.org/www-project-enterprise-security-api/) made by @kwwall 

> But most (perhaps 90% or more) of the ESAPI use which I have observed was solely limited to using ESAPI’s Encoder to remediate XSS vulnerabilities. If that is all that you intend to do with ESAPI, steer clear and use OWASP Java Encoder Project instead

https://owasp.org/www-project-java-encoder

This may be another option for you to investigate.

**Comment by jeremiahjstacey** [comment]
@kwwall , @xeno6696  -- I think we all agree there is no bug in the current Java Logging implementation.  There is potentially a feature request to allow more isolate control of what data gets logged by the ESAPI baseline.  Is that a feature we would like to capture for future iterations of the baseline?

If more support is needed on this issue  I think it would be best continued in a different forum.  I believe that the ESAPI-Users Mailing list would be the recommended arena to continue the configuration discussion, if needed. 
> esapi-project-users@owasp.org 

@kwwall @xeno6696 , can you confirm that is correct forum to continue this discussion?


I recommend closing this issue.

**Comment by kwwall** [comment]
I will be closing this next ESAPI release (tentatively 2.2.3.0, which I am hoping will be completed in a few days max), but I am leaving it open to document that in addition to closing it, I am going to be removing various deprecated methods in ESAPI (that have been marked deprecated for more than 2 years and most which never did anything useful). I probably will create a separate PR just for this since it touches several files.

**Comment by kwwall** [comment]
Closed via PR #603
